testing/unittests/metrics/PeriodicSinkTests.cpp

```diff
@@ -0,0 +1,123 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef _USE_CPPUNIT
+
+#include <cppunit/TestFixture.h>
+#include "unittests.hpp"
+#include <algorithm>
+
+#include "jptree.hpp"
+#include "jmetrics.hpp"
+
+using namespace hpccMetrics;
+
+class PeriodicTestSink : public PeriodicMetricSink
+{
+public:
+    explicit PeriodicTestSink(const char *name, const IPropertyTree *pSettingsTree) :
+        PeriodicMetricSink(name, "test", pSettingsTree) { }
+
+    ~PeriodicTestSink() = default;
+
+protected:
+    virtual void prepareToStartCollecting() override
+    {
+        prepareCalled = true;
+        numCollections = 0;
+    }
+
+    virtual void collectingHasStopped() override
+    {
+        stopCollectionNotificationCalled = true;
+    }
+
+    void doCollection() override
+    {
+        numCollections++;
+    }
+
+public:
+    bool prepareCalled = false;
+    bool stopCollectionNotificationCalled = false;
+    int numCollections = 0;
+};
+
+
+MetricsReporter periodicSinkTestReporter;
+
+const char *periodicSinkSettingsTestYml = R"!!(period: 2
+)!!";
+
+int period = 2;
+
+PeriodicTestSink *pPeriodicTestSink = nullptr;
+
+class PeriodicSinkTests : public CppUnit::TestFixture
+{
+    public:
+        PeriodicSinkTests()
+        {
+            if (pPeriodicTestSink == nullptr)
+            {
+                //
+                // Load the settings then set the period using the global var
+                Owned<IPropertyTree> pSettings = createPTreeFromYAMLString(periodicSinkSettingsTestYml, ipt_none, ptr_ignoreWhiteSpace, nullptr);
+                pSettings->setPropInt("@period", period);
+                pPeriodicTestSink = new PeriodicTestSink("periodic_test_sink", pSettings);
```

Comment:
This will leak, although it will probably avoid being reported as a leak because it is assigned as a static.

Same comments about making the MetricsReporter a member.

Reply:
Moved.